cirq-core/cirq/experiments/random_quantum_circuit_generation.py

```diff
@@ -646,23 +650,34 @@ def __init__(
         self.single_qubit_gates = single_qubit_gates
         self.prng = prng
 
-    def new_layer(self, previous_single_qubit_layer: 'cirq.Moment') -> 'cirq.Moment':
-        def random_gate(qubit: 'cirq.Qid') -> 'cirq.Gate':
-            excluded_op = previous_single_qubit_layer.operation_at(qubit)
-            excluded_gate = excluded_op.gate if excluded_op is not None else None
-            g = self.single_qubit_gates[self.prng.randint(0, len(self.single_qubit_gates))]
-            while g == excluded_gate:
-                g = self.single_qubit_gates[self.prng.randint(0, len(self.single_qubit_gates))]
-            return g
+    def new_layer(
+        self, previous_single_qubit_layer: Dict['cirq.Qid', int]
+    ) -> Dict['cirq.Qid', int]:
+        def random_gate(qubit: 'cirq.Qid') -> int:
+            if qubit not in previous_single_qubit_layer:
+                i = self.prng.randint(0, len(self.single_qubit_gates))
+            else:
+                i = self.prng.randint(0, len(self.single_qubit_gates) - 1)
+                if i >= previous_single_qubit_layer[qubit]:
+                    i += 1
+            return i
 
-        return ops.Moment(random_gate(q).on(q) for q in self.qubits)
+        return {q: random_gate(q) for q in self.qubits}
+
+    def to_moment(self, layer: Dict['cirq.Qid', int]):
+        return ops.Moment(self.single_qubit_gates[layer[q]].on(q) for q in layer.keys())
 
 
 class _FixedSingleQubitLayerFactory:
     def __init__(self, fixed_single_qubit_layer: Dict['cirq.Qid', 'cirq.Gate']) -> None:
         self.fixed_single_qubit_layer = fixed_single_qubit_layer
 
-    def new_layer(self, previous_single_qubit_layer: 'cirq.Moment') -> 'cirq.Moment':
+    def new_layer(
+        self, previous_single_qubit_layer: Dict['cirq.Qid', int]
+    ) -> Dict['cirq.Qid', int]:
+        pass
```

Comment:
is this right? why are we passing here?

Reply:
Yes, the to_moment ignores the input layer, so new_layer doesn't have to generate anything. I'll change the types to Any/None to make it less confusing. (Granted it's still a little confusing, but IDK if it's worth making a base class and generics just for this).

Reply:
updating the type annotations would help, yes